trolldb/database/mongodb.py

```diff
@@ -0,0 +1,201 @@
+"""
+The module which handles database CRUD operations for MongoDB. It is based on
+`PyMongo <https://github.com/mongodb/mongo-python-driver>`_ and `motor <https://github.com/mongodb/motor>`_.
+"""
+
+import errno
+import sys
+from contextlib import asynccontextmanager
+from typing import Any, Coroutine
+
+from loguru import logger
+from motor.motor_asyncio import (
+    AsyncIOMotorClient,
+    AsyncIOMotorDatabase,
+    AsyncIOMotorCollection,
+    AsyncIOMotorCommandCursor,
+    AsyncIOMotorCursor
+)
+from pydantic import validate_call
+from pymongo.collection import _DocumentType
+from pymongo.errors import (
+    ConnectionFailure,
+    ServerSelectionTimeoutError)
+
+from config.config import DatabaseConfig
+
+
+class MongoDB:
+    """
+    A wrapper class around the `motor async driver <https://www.mongodb.com/docs/drivers/motor/>`_ for Mongo DB with
+    convenience methods tailored to our specific needs. As such, most of the methods return coroutines whose results
+    need to be awaited.
+
+    Note:
+        This class is not meant to be instantiated! That's why all the methods in this class are decorated with
+        ``@classmethods``. This choice has been made to guarantee optimal performance, i.e. for each running process
+        there must be only a single motor client to handle all database operations. Having different clients which are
+        constantly opened/closed degrades the performance. The expected usage is that we open a client in the beginning
+        of the program and keep it open until the program finishes. It is okay to reopen/close the client for testing
+        purposes when isolation is needed.
+
+    Note:
+        The main difference between this wrapper class and the original motor driver class is that we attempt to access
+        the database and collections during the initialization to see if we succeed or fail. This is contrary to the
+        behaviour of the motor driver which simply creates a client object and does not attempt to access the database
+        until some time later when an actual operation is performed on the database. This behaviour is not desired for
+        us, we would like to fail early!
+    """
+
+    __client: AsyncIOMotorClient | None = None
+    __main_collection: AsyncIOMotorCollection = None
+    __main_database: AsyncIOMotorDatabase = None
+
+    default_database_names = ["admin", "config", "local"]
+    """
+    MongoDB creates these databases by default for self usage. 
+    """
+
+    @classmethod
+    async def initialize(cls, database_config: DatabaseConfig):
+        """
+        Initializes the motor client. Note that this method has to be awaited!
```

Comment:
I am not sure how this can be done properly without compromising the performance. I think we should look into the implementation details and use cases together.

Reply:
As discussed, we keep the current implementation.